src/atdgen_codec_runtime.ml

```diff
@@ -218,22 +353,22 @@ struct
     else
       raise @@ DecodeError ("Expected array, got " ^ (Js.Json.stringify x))
 
-  let enum l =
-    let constr0 x =
-      let s = string x in
+  let enum l json =
+    let constr0 j = let s = string j in `Constr0 s in
+    let constr j = let p = pair string (fun x -> x) j in `Constr p in
+    match either constr0 constr json with
+    | `Constr0 s -> with_segment s (fun () ->
       match List.assoc s l with
       | exception Not_found -> raise @@ DecodeError (sprintf "unknown constructor %S" s)
       | `Single a -> a
       | `Decode _ -> raise @@ DecodeError (sprintf "constructor %S expects arguments" s)
-    in
-    let constr x =
-      let (s,args) = pair string (fun x -> x) x in
+      ) ()
+    | `Constr (s, args) -> with_segment s (fun () ->
       match List.assoc s l with
       | exception Not_found -> raise @@ DecodeError (sprintf "unknown constructor %S" s)
       | `Single _ -> raise @@ DecodeError (sprintf "constructor %S doesn't expect arguments" s)
-      | `Decode d -> decode d args
-    in
-    either constr0 constr
+      | `Decode d -> decode' d args
```

Comment:
not sure to understand why you don't want to catch the exception here (by using decode instead of decode')

Reply:
Yes, exactly. The `decode` should not be used inside this module - it is the function for an external/public use only. All the exceptions handled & transformed on the way using `with_segment` until the top-most `decode` where it is finally transformed to the public `DecodeError`. This approach in fact permits to have the path in the final exception